### ISSUE TITLE
Split DNS on macOS only fully works with Override local DNS set

### ISSUE THREAD

**Issue by mdeeks** [tailscale/tailscale#4413]
### What is the issue?

DNS queries with command line tools that use resolv.conf do not work with Split DNS when Override local DNS disabled on the Global Name Servers setting
<img width="705" alt="image" src="https://user-images.githubusercontent.com/2966467/163523822-8b47d277-f6af-4e08-9cdf-6cf1be5ab129.png">

`scutil --dns` has the correct settings, but /etc/resolv.conf only points to the resolver that DHCP provides me (1.1.1.1 in my case). So things like web browsers and desktop apps do work fine, but some CLI tools that still rely on resolv.conf do not.

```
❯ cat /etc/resolv.conf
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
search <redacted>
nameserver 1.1.1.1
nameserver 1.0.0.1
```

### Steps to reproduce

* Set up Split DNS for a domain name that is a private zone (e.g. www.yourprivatezone.com)
* * In our case we have a private Route53 zone and set split DNS to point at the AWS resolver in the VPC (e.g. 172.30.0.2)

Broken case:
* Disable Override local DNS
* * /etc/resolv.conf now points to `nameserver 1.1.1.1` provided by DHCP on my local network
* Run `dig www.yourprivatezone.com` and see it fail to find the record

Working case:
* Enable Override local DNS
* * /etc/resolv.conf now points to `nameserver 100.100.100.100`
* Run `dig www.yourprivatezone.com` and see it succeed 

### OS

macOS

### OS version

macOS Monterey 12.2.1

### Tailscale version

1.22.1

**Comment by DentonGentry** [comment]
To set expectations: tailscale doesn't generate the /etc/resolv.conf file, macOS does. We've tried to resolve this by varying the DNS settings pushed to macOS but have been unable to so far. macOS native apps using the same framework as dscacheutil work, but Unix-ish tools reading /etc/resolv.conf get an incomplete set of nameservers.

**Comment by danderson** [comment]
Yeah, that's a known issue, and we can't do anything about it due to macOS's limitations.

macOS's main DNS resolution path is through a system daemon, which understands split DNS and does a bunch of nice stuff. But all the coreutils-style tools that macOS ships as the base system only know to look up /etc/resolv.conf. So, the smart daemon that most stuff uses also writes out a resolv.conf - but because resolv.conf doesn't know split DNS, macOS only populates resolv.conf with whatever the "default" resolver is, ignoring all the split DNS settings.

A better implementation from macOS would have been to mimic what resolved does on linux, and point resolv.conf to a localhost DNS forwarder that's capable of doing all the split stuff. Unfortunately, that's not what Apple's implemented. Due to various other limitations of the NetworkExtension and OS DNS APIs, we also cannot provide that functionality ourselves (macOS won't let us touch resolv.conf, and we can't read out the full DNS config from the OS to mimic resolved's behavior anyway).

As you correctly worked out, if you install a full default-replacing DNS configuration, that leads to consistent behavior with macOS-unaware CLI tools, because then we can point the entire OS at tailscaled's localhost resolver and do all the split stuff ourselves, ignoring the limitations of macOS. But we can't do that if tailscaled doesn't fully own the system DNS config :(

I'm not much of an iDevice user myself, but if macOS has a mechanism to send feedback to Apple, asking them to implement resolv.conf management using a local DNS forwarder daemon would fix this issue OS-wide.

**Comment by mdeeks** [comment]
> and we can't read out the full DNS config from the OS to mimic resolved's behavior anyway

Ahh I see now. I was going to suggest just taking what my OS already has but it sounds like you can't even access that. Thats a bummer but it doesn't seem like there is much you all can do without Apple making a change to the OS. It is really surprising they don't offer that via APIs though.

Granted I know zero about macOS development but does this interface not offer a way to find it via loadFromPreferences()? 
https://developer.apple.com/documentation/networkextension/nednssettingsmanager
https://developer.apple.com/documentation/networkextension/nednssettings

> if macOS has a mechanism to send feedback to Apple, asking them to implement resolv.conf management using a local DNS forwarder daemon would fix this issue OS-wide.

I don't even know where to begin to ask that. That really seems like the only solution though.

We will leave Override local DNS enabled. It isn't a big deal except for a very small subset of users who want to use a resolver on their own home network.

**Comment by digitaldan** [comment]
@danderson i was thinking a good compromise / workaround might be to allow setting DNS override on a device by device basis in the portal.  For example, i don't wan't to blanket override DNS for all users/machines in my system, but it would be nice to set if for specific devices, especially mobile IOS devices.  I am aware that Desktop and CLI clients allow you to "use Tailscale DNS Settings", but thats a pain to make users disable, and could still be useful on top of a portal setting.  Plus  mobile does not have this feature in any case.  Just my 2 cents.

**Comment by ilf** [comment]
I just ran into this issue. I was shocked to learn that the Tailscale of my organisation forces `override_local_dns: true` on all users and all plattforms, because of this macOS issue.

Since changing macOS behavior seems unlikely, I would like other ways to work around this. How about a setting like `override_local_dns_on_macos: true` that only does this on macOS, but doesn't hijack the system DNS of platforms behaving correctly?
